### DISCUSSION TITLE
Non-destructive request header access

### DISCUSSION THREAD

**Discussion by kellytk** [tokio-rs/axum#467] [Q&A] [answered]
I encountered an issue when reordering a handler's extractors that caused form submission data to cease posting as expected.

The offending extractor was HeaderMap and moving it to the end of the extractors solves the problem as the [docs](https://docs.rs/axum/0.3.0/axum/extract/index.html#applying-multiple-extractors) advise.  I dislike having bugs emerge due to argument order changes and am seeking a permanent solution.

What options are available for accessing request header data without consuming it and causing other extractors to fail?

**Comment by davidpdrsn** [comment, marked as answer]
The design of extractors doesn't allow for this unfortunately. Extractors that consume the request body have the same problems, such as `String`, `Bytes`, and `Json`. So even if we made `HeaderMap` clone all the headers (which I don't think we should) body extractors would still have the issue.

I recommend you write own `HeaderMapClone` extractor that clones the headers and leaves the original ones in place for other extractors.

**Reply by kellytk** [reply]
Another option is to create cloning counterparts to the current extractors so users can choose as appropriate.  Would you be interested in pull requests?

**Reply by davidpdrsn** [reply]
If we did that we would get into trouble when creating other extractors that wrap one that mutates the request. Then we would also have to have two versions of the wrapper: one that mutates and another that clones. Feels like we risk getting an exploration of extractors suddenly. Admittedly a small risk but still something to consider.

It also doesn't solve the issues with request bodies. That would require buffering the body and keeping it in memory (not even sure where you'd store it, I guess in request extensions). That is expensive and impossible if the body is a long running stream.

So no I don't think we should have such extractors in axum, however you're of course free to make a crate that provides them. We could link to it from [the community showcase](https://github.com/tokio-rs/axum/blob/main/ECOSYSTEM.md#project-showcase).

**Reply by davidpdrsn** [reply]
Another thing worth noting is that this only applies to headers, as you cannot clone request extensions.

**Reply by davidpdrsn** [reply]
The idea solution would be if we could make this cause type errors instead of runtime errors. I spent a bunch of time on this before the initial release and wasn't able to find a solution 😞 You're welcome to give it a shot :)

**Reply by kellytk** [reply]
Thank you for the helpful information.  I'll ponder those options.  Separately but related, how can it be determined if an extractor mutates the request, and if so, whether it mutates the headers and/or body?  I'd like to better comprehend why I could use the Form extractor before HeaderMap, but not the reverse.

**Reply by davidpdrsn** [reply]
There isn't a better way than looking at the implementation unfortunately.

All extractors in axum also have rejections that aim to be clear about what the error was. For example `Form`'s rejection is [`FormRejection`](https://docs.rs/axum/0.3.0/axum/extract/rejection/enum.FormRejection.html) which has a variant called `HeadersAlreadyExtracted`. That tells you `Form` wont work if all headers have been extracted first. It needs the headers to [look at the `Content-Type`](https://github.com/tokio-rs/axum/blob/main/axum/src/extract/form.rs#L63). Of course this wont always work for extractors provided by other libraries so it should be documented or otherwise clear from the types.

In general I would also recommend extracting only the headers you need via [`TypedHeader`](https://docs.rs/axum/0.3.0/axum/extract/struct.TypedHeader.html) without removing all headers with `HeaderMap`.

**Reply by kellytk** [reply]
Thanks once again.  Would you accept a PR adding a note to https://docs.rs/axum/0.3.0/axum/extract/struct.TypedHeader.html that it is recommended over HeaderMap wherever possible?

**Reply by davidpdrsn** [reply]
Definitely!

**Reply by kellytk** [reply]
[Done](https://github.com/tokio-rs/axum/pull/476).

In support of your eventual ideal solution, and as a learning resource and reference to users, what's your opinion of adding a table to https://docs.rs/axum/0.3.0/axum/extract/index.html that details the requirements and mutations for each extractor?  For example, the `Form` extractor's requirement is the Content-Type header, and the `HeaderMap`'s mutation takes the headers.

With the search query https://docs.rs/axum/0.3.0/axum/extract/index.html?search=headers I've already found how I can ferret out much of the information despite being new to axum.

**Reply by davidpdrsn** [reply]
I feel that might be overkill and tedious to maintain 🤷‍♂️ It's only relevant to consider when using `HeaderMap` and body extractors and those are already documented.
